include/zephyr/bindesc.h

```diff
@@ -0,0 +1,405 @@
+		.len = (uint16_t)sizeof(uint32_t),	\
+		.data = UINT_TO_BYTE_ARRAY(value),	\
+	}
+
+/**
+ * @brief Define a binary descriptor of type bytes.
+ *
+ * @details
+ * Define a uint8_t array that is registered in the binary descriptor header.
+ * The defined array can be accessed using @ref BINDESC_GET_BYTES.
+ * The value should be given as an array literal, wrapped in parentheses, for
+ * example:
+ *
+ *     DEFINE_BINDESC_BYTES(name, id, ({1, 2, 3, 4}));
+ *
+ * @note The defined array is not static, so its name must not collide with
+ * any other symbol in the executable.
+ *
+ * @param name Name of the descriptor
+ * @param id Unique ID of the descriptor
+ * @param value A uint8_t array as data for the descriptor
+ */
+#define DEFINE_BINDESC_BYTES(name, id, value)				\
+	DEFINE_BINDESC_ENTRY(name) = {					\
+		.tag = BINDESC_TAG(BYTES, id),				\
+		.len = (uint16_t)sizeof((uint8_t [])__DEBRACKET value),	\
+		.data = __DEBRACKET value,				\
+	}
+
+/**
+ * @brief Get the value of a string binary descriptor
+ *
+ * @details
+ * Get the value of a string binary descriptor, previously defined by
+ * DEFINE_BINDESC_STR.
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_STR(name) BINDESC_NAME(name).data
+
+/**
+ * @brief Get the value of a uint binary descriptor
+ *
+ * @details
+ * Get the value of a uint binary descriptor, previously defined by
+ * DEFINE_BINDESC_UINT.
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_UINT(name) *(uint32_t *)&(BINDESC_NAME(name).data)
+
+/**
+ * @brief Get the value of a bytes binary descriptor
+ *
+ * @details
+ * Get the value of a string binary descriptor, previously defined by
+ * DEFINE_BINDESC_BYTES. The returned value can be accessed as an array:
+ *
+ *     for (size_t i = 0; i < BINDESC_GET_SIZE(name); i++)
+ *         BINDESC_GET_BYTES(name)[i];
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_BYTES(name) BINDESC_NAME(name).data
+
+/**
+ * @brief Get the size of a binary descriptor
+ *
+ * @details
+ * Get the size of a binary descriptor. This is particularly useful for
+ * bytes binary descriptors where there's no null terminator.
+ *
+ * @param name Name of the descriptor
+ */
+#define BINDESC_GET_SIZE(name) BINDESC_NAME(name).len
+
+/**
+ * @}
+ */
+
+#if !IS_ENABLED(_LINKER)
+
+/*
+ * An entry of the binary descriptor header. Each descriptor is
+ * described by one of these entries.
+ */
+struct bindesc_entry {
+	/** Tag of the entry */
+	uint16_t tag;
+	/** Length of the descriptor data */
+	uint16_t len;
+	/** Value of the entry. This is either an integer or a string */
+	uint8_t data[];
+} __packed;
+
+/*
+ * We're assuming that `struct bindesc_entry` has a specific layout in
+ * memory, so it's worth making sure that the layout is really what we
+ * think it is. If these assertions fail for your toolchain/platform,
+ * please open a bug report.
+ */
+BUILD_ASSERT(offsetof(struct bindesc_entry, tag) == 0, "Incorrect memory layout");
```

Comment:
These lines could be replaced with:
`BUILD_ASSERT(sizeof(struct bindesc_entry) == 4))`
no?

Reply:
The entry may contain a string which is of arbitrary size

Reply:
not `struct bindesc_entry`, that only contains an empty array.

Reply:
Even if I'm right I don't mind if you leave this as-is, it's not blocking.